### PR TITLE
feat: add flyout as a new component

### PR DESCRIPTION
**Pull request by chrisdholt**
# Description

Adds a new styled component “flyout” to the MSFT package.

## Issue type checklist

- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

## Process & policy checklist

- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast-dna/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

File: packages/fast-components-react-msft/src/flyout/flyout.tsx [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        verticalThreshold: void 0,
+        viewport: void 0,
+        width: void 0,
+    };
+
+    /**
+     * Renders the component
+     */
+    public render(): React.ReactElement<HTMLDivElement> {
+        // const managedClassMap: ViewportPositionerClassNameContract = this.generateManagedClassNames();
```
Comment (nicholasrice) [inline packages/fast-components-react-msft/src/flyout/flyout.tsx]: Probably don't need this.

File: packages/fast-components-react-msft/src/flyout/flyout.tsx [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                this.props.managedClasses,
+                "flyout__verticalInset",
+                ""
+            ),
+        };
+    }
+
+    private handleWindowClick = (event: MouseEvent): void => {
+        if (typeof this.props.onDismiss === "function" && !this.props.disabled) {
+            this.props.onDismiss(event);
```
Comment (nicholasrice) [inline packages/fast-components-react-msft/src/flyout/flyout.tsx]: Whats the story if the user clicks *inside* the flyout? Do we still want to dismiss the flyout?

File: packages/fast-components-react-msft/src/flyout/flyout.tsx [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        verticalPositioningMode: FlyoutAxisPositioningMode.adjacent,
+        defaultVerticalPosition: FlyoutVerticalPosition.bottom,
+        horizontalAlwaysInView: false,
+        verticalAlwaysInView: false,
+        fixedAfterInitialPlacement: false,
+        height: "128px",
+        width: "280px",
+    };
+
+    protected handledProps: HandledProps<FlyoutHandledProps> = {
```
Comment (nicholasrice) [inline packages/fast-components-react-msft/src/flyout/flyout.tsx]: if you change this to `HandledProps<Required<FlyoutHandledProps>>` we won't run into the case where this doesn't cause a build error if optional props aren't in `handledProps`

File: packages/fast-components-react-msft/src/flyout/flyout.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        if (!this.state.visible) {
+            return;
+        }
+
+        const { anchor, children, disabled, ...props }: FlyoutProps = this.props;
+
+        return (
+            <Flyout
+                anchor={this.anchor.current}
+                disabled={!this.state.visible}
```
Comment (nicholasrice) [inline packages/fast-components-react-msft/src/flyout/flyout.stories.tsx]: "disabled" isn't an intuitive name to me to convey whether the flyout is visible or not - should we deviate from the viewport positioner API here?

File: packages/fast-components-react-msft/src/flyout/flyout.tsx (hunk 2 above)
Comment (chrisdholt) [reply packages/fast-components-react-msft/src/flyout/flyout.tsx]: touché - good catch.

File: packages/fast-components-react-msft/src/flyout/flyout.stories.tsx (hunk 4 above)
Comment (chrisdholt) [reply packages/fast-components-react-msft/src/flyout/flyout.stories.tsx]: I think we should - visible makes more sense to me. I’ll update the contract.

File: packages/fast-components-react-msft/src/flyout/README.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# Flyout
+The *flyout* component accepts children and follows the functional pattern of a non-modal dialog (non-modal dialogs allow users the ability to continue interacting with the page). The width and height of the flyout can be set by props. If no props are passed to set width or height, the dialog width will default to 280px and the height will default to 128px.
```
Comment (nalogiudice) [inline packages/fast-components-react-msft/src/flyout/README.md]: nit: "...props are passed to set width or height, the **dialog** width will..." **dialog** should be **flyout** here to avoid confusion I think

File: packages/fast-components-react-msft/src/flyout/flyout.tsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+    /**
+     * Renders the component
+     */
+    public render(): React.ReactElement<HTMLDivElement> {
+        return (
+            <ViewportPositioner
+                {...this.unhandledProps()}
+                role="dialog"
+                aria-label={this.props.label}
```
Comment (nalogiudice) [inline packages/fast-components-react-msft/src/flyout/flyout.tsx]: These props (aria-label, aria-labelledby, aria-describedby) are optional, but you aren't checking to see if they exist before using them?

File: packages/fast-components-react-msft/src/flyout/flyout.tsx (hunk 2 above)
Comment (nalogiudice) [reply packages/fast-components-react-msft/src/flyout/flyout.tsx]: ^ just to be clear no, correct? The user should be able to interact with content inside the flyout - that's the point of a dialog-type flyout

Comment (chrisdholt) [reply packages/fast-components-react-msft/src/flyout/flyout.tsx]: correct.

File: packages/fast-components-react-msft/src/flyout/flyout.schema.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
+export default {
+    $schema: "http://json-schema.org/schema#",
+    title: "Flyout",
+    description: "A flyout component's schema definition.",
+    type: "object",
+    id: "@microsoft/fast-components-react-msft/flyout",
+    properties: {
```
Comment (eljefe223) [inline packages/fast-components-react-msft/src/flyout/flyout.schema.ts]: does this need a `formPluginId`?

File: packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.tsx [hunk 8]
```diff
@@ -327,6 +327,9 @@ class ViewportPositioner extends Foundation<
      *  gets the CSS classes to be programmatically applied to the component
      */
     private getPositioningStyles = (): React.CSSProperties => {
+        // Check if there is already a style object being passed as props
+        const styleProps: React.CSSProperties = get(this.props, "style");
```
Comment (janechu) [inline packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.tsx]: It seems to me that since this change is happening in `@microsoft/fast-components-react-base` it will add to the changelog a minor bump and the item "add flyout as a new component". This doesn't seem desirable since this should be a change to msft packages only. Can we split the changes here into another PR?

File: packages/fast-components-react-msft/src/flyout/flyout.schema.ts (hunk 7 above)
Comment (janechu) [reply packages/fast-components-react-msft/src/flyout/flyout.schema.ts]: I've been adding them to schemas for the tooling, so I would say both a `formPluginId` and a `pluginId` would be great

File: packages/fast-components-react-msft/src/flyout/flyout.tsx [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+                    ViewportPositionerHorizontalPosition[
+                        FlyoutHorizontalPosition[this.props.defaultHorizontalPosition]
+                    ]
+                }
+                defaultVerticalPosition={
+                    ViewportPositionerVerticalPosition[
+                        FlyoutVerticalPosition[this.props.defaultVerticalPosition]
+                    ]
+                }
+                disabled={!this.props.visible}
```
Comment (scomea) [inline packages/fast-components-react-msft/src/flyout/flyout.tsx]: Does this make sense?  When the positioner is disabled it behaves like a regular div, the positioning logic does not run but it is visible.

File: packages/fast-components-react-msft/src/flyout/flyout.tsx (hunk 6 above)
Comment (chrisdholt) [reply packages/fast-components-react-msft/src/flyout/flyout.tsx]: Sure - if these are null or undefined, they won’t get passed

File: packages/fast-components-react-msft/src/flyout/flyout.tsx [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+                    ViewportPositionerHorizontalPosition[
+                        FlyoutHorizontalPosition[this.props.defaultHorizontalPosition]
+                    ]
+                }
+                defaultVerticalPosition={
+                    ViewportPositionerVerticalPosition[
+                        FlyoutVerticalPosition[this.props.defaultVerticalPosition]
+                    ]
+                }
+                disabled={false}
```
Comment (scomea) [inline packages/fast-components-react-msft/src/flyout/flyout.tsx]: This is the default, so not required

Comment (chrisdholt) [reply packages/fast-components-react-msft/src/flyout/flyout.tsx]: thx - I’m leaving here as explicit for now
